Rules/UseUsingScopeModifierInNewRunspaces.cs

```diff
@@ -0,0 +1,430 @@
+
+        private readonly Dictionary<string, List<VariableExpressionAst>> _varsDeclaredPerSession = new Dictionary<string, List<VariableExpressionAst>>();
+        
+        private readonly List<DiagnosticRecord> _diagnosticAccumulator;
+        
+        private readonly UseUsingScopeModifierInNewRunspaces _rule;
+        
+        private readonly string _analyzedFilePath;
+        
+        public SyntaxCompatibilityVisitor(UseUsingScopeModifierInNewRunspaces rule, string analyzedScriptPath)
+        {
+            _diagnosticAccumulator = new List<DiagnosticRecord>();
+            _rule = rule;
+            _analyzedFilePath = analyzedScriptPath;
+        }
+
+        /// <summary>
+        /// GetDiagnosticRecords: Retrieves all Diagnostic Records that were generated during visiting
+        /// </summary>
+        public IEnumerable<DiagnosticRecord> GetDiagnosticRecords()
+        {
+            return _diagnosticAccumulator;
+        }
+
+        /// <summary>
+        /// VisitScriptBlockExpression: When a ScriptBlockExpression is visited, see if it belongs to a command that needs its variables
+        /// prefixed with the 'Using' scope modifier. If so, analyze the block and generate diagnostic records for variables where it is missing.
+        /// </summary>
+        /// <param name="scriptBlockExpressionAst"></param>
+        /// <returns>
+        /// AstVisitAction.Continue or AstVisitAction.SkipChildren, depending on what we found. Diagnostic records are saved in `_diagnosticAccumulator`.
+        /// </returns>
+        public override AstVisitAction VisitScriptBlockExpression(ScriptBlockExpressionAst scriptBlockExpressionAst)
+        {
+            if (!(scriptBlockExpressionAst.Parent is CommandAst commandAst))
+            {
+                return AstVisitAction.Continue;
+            }
+
+            var cmdName = commandAst.GetCommandName();
+            if (cmdName == null)
+            {
+                // Skip for situations where command name cannot be resolved like `& $commandName -ComputerName -ScriptBlock { $foo }` 
+                return AstVisitAction.Continue;
+            }
+            
+            var scriptBlockParameterAst = commandAst.CommandElements[
+                    commandAst.CommandElements.IndexOf(scriptBlockExpressionAst) - 1] as CommandParameterAst;
+
+            if (IsInlineScriptBlock(cmdName) || 
+                IsJobScriptBlock(cmdName, scriptBlockParameterAst) ||
+                IsForeachScriptBlock(cmdName, scriptBlockParameterAst) ||
+                IsInvokeCommandComputerScriptBlock(cmdName, commandAst) ||
+                IsDSCScriptResource(cmdName, commandAst))
+            {
+                AnalyzeScriptBlock(scriptBlockExpressionAst);
+                return AstVisitAction.SkipChildren;
+            }
+
+            if (IsInvokeCommandSessionScriptBlock(cmdName, commandAst))
+            {
+                var sessionName = GetSessionName(commandAst);

```

Comment:
The type of the LHS isn't indicated on the RHS here. Would it be possible to use an explicit type signature? (Here and with other `var` uses resulting from a method call)

Reply:
Yes, no problem, done. Maybe there is a setting I can change to make Visual Studio stop complaining. Visual Studio keeps pointing out to me that I can change these to `var`, but I certainly see your point for readability.

Reply:
Yeah VS seems to have some configurations that are overzealous. From my perspective, I spend a lot of time reading code, and often not in an IDE. That means being able to infer the type of something on the same line it's defined is valuable, so my general rule is that `var` should be used only when the type is expressed on the right hand side:

- `var x = new Thing();`
- `var x = y as Thing;`
- `var x = JsonConvert.Convert<Thing>(str);`
- The iffy `var x = Thing.Create();`

But the antipatterns in my mind are things like:

```csharp
var x = w.Method();
var y = x.Property;
var z = y.AnotherMethod(); // If we got x wrong, the compile error might only show here
```

or

```csharp
var x = 7; // int has the same number of characters as var...
```